plugins/woocommerce/src/Internal/DataStores/CustomMetaDataStore.php

```diff
@@ -232,4 +232,35 @@ public function get_metadata_by_key( &$object, string $meta_key ) {
 		return $meta;
 	}
 
+	/**
+	 * Returns distinct meta keys in use.
+	 *
+	 * @since 8.8.0
+	 *
+	 * @param int    $limit           Maximum number of meta keys to return. Defaults to 100.
+	 * @param string $order           Order to use for the results. Either 'ASC' or 'DESC'. Defaults to 'ASC'.
+	 * @param bool   $include_private Whether to include private meta keys in the results. Defaults to FALSE.
+	 * @return string[]
+	 */
+	public function get_meta_keys( $limit = 100, $order = 'ASC', $include_private = false ) {
+		global $wpdb;
+
+		$db_info = $this->get_db_info();
+
+		$query = "SELECT DISTINCT meta_key FROM {$db_info['table']} ";
+
+		if ( ! $include_private ) {
+			$query .= $wpdb->prepare( 'WHERE meta_key NOT LIKE %s ', $wpdb->esc_like( '_' ) . '%' );
+		}
+
+		$order  = in_array( $order, array( 'ASC', 'DESC' ), true ) ? $order : 'ASC';
+		$query .= 'ORDER BY meta_key ' . $order . ' ';
+
+		if ( $limit ) {
+			$query .= $wpdb->prepare( 'LIMIT %d ', $limit );
+		}
+
+		return $wpdb->get_col( $query ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared -- $query is prepared.
```

Comment:
🤔 Seems like it might be worth caching the result of this query so it doesn't have to run on every Edit Order page load...

Reply:
I'm not saying we shouldn't, but this is how it works in WP too. I guess if we cache the results we might omit keys that were added (not just in the editor itself but programmatically). If you feel this is a blocker, I'd be happy to add some caching, but I'm not entirely sure yet how we would invalidate it (other than it being a transient). Thoughts?

Reply:
Not a blocker. If WP Core does the same, that's pretty battle-tested, so probably not necessary for us to go above and beyond.

> I guess if we cache the results we might omit keys that were added (not just in the editor itself but programmatically).

Yeah, this is a good point. There are some actions we could hook into when postmeta gets changed, where we could then delete a transient, but you're probably right that there would be edge cases that would be hard to catch.

Reply:
Hey @jorgeatorres @coreymckrill , this function ends up creating a query like

```
SELECT
  DISTINCT meta_key
FROM
  wp_wc_orders_meta
WHERE
  meta_key NOT LIKE '\\_%'
ORDER BY
  meta_key ASC
LIMIT
  30
```

In our case, we have 70m+ records and the table is 20Gb+, so the query takes 5-6 minutes to run and breaks the order edit and create pages on admin.

#46476 resolves the issue, but it won't go live until 8.9.

This issue got us into a bit of a panic when we upgraded to 8.8.2 and it's not something we could reproduce on staging (no data), so in the future, it's worth testing wildcard queries on larger tables.

In an ideal world, Woo would also have unit tests running specifically for large stores, but I understand this is unlikely to happen.

Reply:
> in the future, it's worth testing wildcard queries on larger tables
> In an ideal world, Woo would also have unit tests running specifically for large stores

Good feedback, thank you @glagonikas